packages/connect-query-core/src/connect-query-key.ts

```diff
@@ -246,5 +255,27 @@ export function createConnectQueryKey<
       );
     }
   }
+  if (
+    params.schema.kind === "rpc" &&
+    "headers" in params &&
+    params.headers !== undefined
+  ) {
+    props.headers = createHeadersKey(params.headers);
+  }
   return ["connect-query", props] as ConnectQueryKey<O>;
 }
+
+/**
+ * Creates a record of headers from a HeadersInit object.
+ */
+function createHeadersKey(headers: HeadersInit): Record<string, string> {
```

Comment:
We have to make this a bit more robust. Consider the following declarations:

```ts
const a: HeadersInit = {
  foo: "a, b",
};

const b: HeadersInit = [
  ["foo", "a"],
  ["foo", "b"],
];

const c: HeadersInit = [
  ["foo", "a"],
  ["Foo", "b"],
];
```

They are different formulations for exactly the same headers, but would result in different keys. The most reliable and simple way to handle headers init is to normalize via `new Headers()`.

I suggest adding to the JSDoc:
> Note that invalid HTTP header names will raise a TypeError, and that the `Set-Cookie` header is not supported.